DataLayout related support.

svvivek opened the pull request:
This is an attempt at capturing DataLayout structurally. The goal of this PR is just to support the following basic things so that this can help build other things which needed this structure/information as a prerequisite.

- Generate datalayout with default settings.
- Parse datalayout from string.
- Generate datalayout string from structure.

I have done basic testing to see if it generated valid layout string and LLVM understands it without any issues. Adding this point as I haven't written any test cases for this.

dannypsnl commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	return &FunctionPointerAlignment{isIndependant: isInd, abiAlignment: abiAl}
+}
+
+type ManglingStyle string
+
+const (
+	ELF        ManglingStyle = "e"
+	Mips       ManglingStyle = "m"
+	MachO      ManglingStyle = "o"
+	WinX86COFF ManglingStyle = "x"
+	WinCOFF    ManglingStyle = "w"
+	XCOFF      ManglingStyle = "a"
+)
```
I would like to see them in a separate file.

dannypsnl replied:
I think a submodule for `datalayout` is ok for me.

svvivek replied:
I initially had data_layout_util.go file with those constants and other utils/helpers/string(parsing/generation) other than the main structure and its methods present in this file. But merged it before sending this PR.

dannypsnl replied:
ok, it's fine for now.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	return &FunctionPointerAlignment{isIndependant: isInd, abiAlignment: abiAl}
+}
+
+type ManglingStyle string
```
Consider moving `ManglingStyle` to the `ir/enum` package. We may consider making an integer enum with `String` methods generated using `stringer`. See https://github.com/llir/llvm/blob/9e62c17a9f75d127c4010b0d242912413c9d73cc/ir/enum/enum.go#L42 for an example.

dannypsnl replied:
I think we should do this later; else a PR depends on another PR will be hard to manage 

And we can do that easier since we are owner

dannypsnl commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	programMemoryAddressSpace uint64                                 // default - 0
+	globalVarAddressSpace     uint64                                 // default - 0
+	allocaAddressSpace        uint64                                 // default - 0
+	ptrSizeAlign              map[uint64]*PointerSizeAlignment       // Key is addressspace
+	intSizeAlign              map[uint64]*IntegerSizeAlignment       // Key is size
+	vecSizeAlign              map[uint64]*VectorSizeAlignment        // Key is size
+	floatSizeAlign            map[uint64]*FloatingPointSizeAlignment // Key is size
```
Does there have any reasons we need to use a pointer to the structure instead of the structure itself?

svvivek replied:
I am not sure I understand the question correctly. 
We need instances of those structures and pointer is the way to represent it right ?

dannypsnl replied:
For example `map[uint64]VectorSizeAlignment` instead of  `map[uint64]*VectorSizeAlignment`

dannypsnl replied:
Go does not require structure with pointer

svvivek replied:
> For example `map[uint64]VectorSizeAlignment` instead of `map[uint64]*VectorSizeAlignment`
> Go does not require structure with pointer

Doing this would mean, they are passed by values when they are given as arguments or returned from functions, thereby changes made to them not getting reflected if not re assigned properly. Whereas, pointers would reflect any changes made to it no matter how an instance is used. I think its better this way.

dannypsnl replied:
Not actually, because you can use & to get a reference whenever you want

svvivek replied:
That can be done, but that is one more thing to include in review to make sure we haven't forgotten to add & wherever required. 

Anyway, I checked for llir/llvm repo and the convention there is use of pointers in case of structs https://github.com/llir/llvm/blob/9e62c17a9f75d127c4010b0d242912413c9d73cc/ir/func.go#L19 so I think this follows the existing pattern.

dannypsnl reviewed (changes requested):
Looks good, just a bit about code construction

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+// DataLayout is a structural representation of the datalayout
+// string from https://llvm.org/docs/LangRef.html#data-layout
+type DataLayout struct {
+	isBigEndian               bool                                   // default - True
```
For consistency with other parts of the llir API, export the fields of `DataLayout` instead of using getters/setters.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	return val
+}
+
+func (dl *DataLayout) SetIsBigEndian(isBigEndian bool) {
```
See comment above, export fields of `DataLayout` and remove getters/setters for consistency with other parts of the llir API.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+
+// NewDataLayout returns an instance of DataLayout with default values
+// taken from https://llvm.org/docs/LangRef.html#data-layout
+func NewDataLayout(os string, arch string) *DataLayout {
```
If possible, we should consider making `arch` an enum.

svvivek replied:
@dannypsnl's comment looks better to me. If you guys can add the required enums there, I can just use it in this PR instead of opening another one.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+		} else if strings.HasPrefix(spec, "F") {
+			addFuncPtrAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "m") {
+			dl.mangling = ManglingStyle(strings.Split(spec, ":")[1])
```
We should add error detection to prevent the library from crashing if `len(strings.Split(spec, ":"))` is less than 2. In general the library should on panic when we have code paths that are not fully implemented (e.g. a switch missing an enum value case branch). But for code that parses arbitrary user input, we should return an error rather than crashing with a panic.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	} else if strings.HasPrefix(val, "n") {
+		typ = "n"
+	} else {
+		panic("Invalid Function pointer alignment type, possible options are \"i\", \"n\".")
```
Adding the full `spec` in error string to make it easier to debug.
```suggestion
		panic(fmt.Errorf(`invalid function pointer alignment type, expected "i" or "n" prefix, got %q`, spec))
```

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+func getUInt64(str string) uint64 {
+	val, err := strconv.ParseUint(str, 10, 64)
+	if err != nil {
+		panic(err.Error())
```
While convenient to use `MustFoo` kind of functions, we should expect errors in arbitrary user input and report errors rather than panic. So, return errors when parsing strings that are expected to be unsigned integers.

In general, libraries should strive to return errors so users of the library may decide how to handle errors, rather than panicking and crashing applications using the library. The exception to this rule is if there are code paths in our `llir` library that we may update to e.g. add a new enum case to cover new code paths.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	dl.nativeIntBitWidths = bitWidths
+}
+
+func (dl *DataLayout) GetLayoutString() string {
```
We can probably change the `GetLayoutString` method name to either `String` to implement `fmt.Stringer`, or perhaps `LLString() string` to implement [ir.LLStringer](https://pkg.go.dev/github.com/llir/llvm/ir#LLStringer).

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+}
+
+func (dl *DataLayout) GetLayoutString() string {
+	layout := ""
```
To handle memory allocation, we could consider using `strings.Builder` to generate the string. See https://github.com/llir/llvm/blob/9e62c17a9f75d127c4010b0d242912413c9d73cc/ir/func.go#L126 for an example.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+}
+
+type PointerSizeAlignment struct { // All sizes are in bits.
+	addressSpace            uint64
```
Same comment as above, export struct fields for consistency with other parts of the `llir` API.

mewmew commented on data_layout.go:
```diff
@@ -0,0 +1,421 @@
+	return floatAlignMap
+}
+
+type AggregateAlignment struct {
```
If possible, add a simple comment for each exported type to describe what it is used for. If there are documentation in the LLVM language reference, then it would be great to add links to those. If not, that's fine.

svvivek replied:
I already added a common comment to the DataLayout struct with the Ref Doc link which has all the details. Since there are no anchor links to each spec property. I don't think a short text would do justice for the information given there and I'll end up copying whole description of each spec here. So I am thinking I'll leave it as it is. If you still prefer adding it, I'll do it but we'll need to track everytime a release happens for changes in main doc and update here appropriately.

mewmew reviewed (commented):
Really excited to see you working on this @svvivek! It's a great start and well needed functionality for `irutil` to have the ability to parse data layout strings. Also happy to see you tackled the generation of data layout strings! This will make it easier to implement round-trip test cases.

Speaking of which, do you think you could try to add a round-trip test case for data layout parsing and string generation? You can look for example strings in the [testdata](https://github.com/llir/testdata) repo. The round trip would have *golden* data layout input strings (extracted from the testdata repo), then parse those strings into `irutil.DataLayout` structs, then invoke `String` (or `GetLayoutString`) to check that the generated string is the same as the input string.

I have given a few initial review comments, just basic review feedback.

As I'm current on vacation I have not had time to go through the PR in detail. I'll head out on a hike in the next few days :)

Still, very happy to see you start work on this! I'd be happy to merge the PR after we've done more thorough review and updated a few things (such as error handling, comments, exported fields vs getters/setters).

Wish you a happy summer.

Cheers,
Robin

svvivek commented:
> The round trip would have _golden_ data layout input strings (extracted from the testdata repo), then parse those strings into `irutil.DataLayout` structs, then invoke `String` (or `GetLayoutString`) to check that the generated string is the same as the input string.

This is how I tested before sending this PR :smile: but with my own data instead of that testdata repo. Only problem with the current structure I see is that since I have maps, ranging over it would be in random order, so direct string comparison won't work, need to think about it.

Regarding other comments : - I'll do what is possible with the time I have got, enjoy the vacation :smile:

mewmew commented:
> This is how I tested before sending this PR smile but with my own data instead of that testdata repo. Only problem with the current structure I see is that since I have maps, ranging over it would be in random order, so direct string comparison won't work, need to think about it.

It would be preferable to make the output deterministic. The way to do this is to first range through the map keys, then sort them (e.g. int or string sort). Then use the sorted list of keys for deterministic output.

svvivek commented:
> > This is how I tested before sending this PR smile but with my own data instead of that testdata repo. Only problem with the current structure I see is that since I have maps, ranging over it would be in random order, so direct string comparison won't work, need to think about it.
> 
> It would be preferable to make the output deterministic. The way to do this is to first range through the map keys, then sort them (e.g. int or string sort). Then use the sorted list of keys for deterministic output.

This is what I have in mind too :+1:

dannypsnl commented:
> > This is how I tested before sending this PR smile but with my own data instead of that testdata repo. Only problem with the current structure I see is that since I have maps, ranging over it would be in random order, so direct string comparison won't work, need to think about it.
> 
> It would be preferable to make the output deterministic. The way to do this is to first range through the map keys, then sort them (e.g. int or string sort). Then use the sorted list of keys for deterministic output.

`reflect.DeepEqual` should be enough?

svvivek committed:
- DataLayout related support.
  - Generate datalayout with default settings.
  - Parse datalayout from string.
  - Generate datalayout string from structure.

svvivek committed:
- PR #4 review comments related changes.
  - Exported fields and removed setters/getters.
  - Used strings.Builder.
  - DataLayout now implements ir.LLStringer.

dannypsnl committed:
better error message

Co-authored-by: Robin <mewmew@users.noreply.github.com>

svvivek committed:
- PR #4 review comments related changes.
  - errors are now returned instead of panic.
  - error prone cases are handled in possible cases.

svvivek committed:
Merge branch 'master' of https://github.com/svvivek/irutil

dannypsnl commented:
@mewmew Do you still have any questions about the PR? If no, I would like to merge it.

mewmew commented:
> @mewmew Do you still have any questions about the PR? If no, I would like to merge it.

I would still like to have the test cases added, since there is a lot of parsing logic in this PR. Having it validated by test cases would be preferable.

re: https://github.com/llir/irutil/pull/4#issuecomment-901865176, @svvivek did the strategy work to make the test cases deterministic?

Cheers,
Robin

svvivek commented:
> re: [#4 (comment)](https://github.com/llir/irutil/pull/4#issuecomment-901865176), @svvivek did the strategy work to make the test cases deterministic?

@mewmew  Unfortunately, I couldn't find time writing test cases yet due to other personal priorities.

Btw, based on the lang ref doc, I added defaults values whereever possible (as these entries might be helpful somewhere) when creating DataLayout struct, so the string I generate from that instance will have those extra spec entries which the layout string from the test data won't have, so direct string comparison will also not be possible, all we can do is create a datalayout instance from test data string, then generate a layout string, create another datalayout from this new string and use DeepEqual as @dannypsnl suggested. This is what I'll do, when I get time for it.

In the meantime, if you guys can add required enums in the other repo as @dannypsnl suggested, I'll update those here too.

mewmew commented:
> @mewmew Unfortunately, I couldn't find time writing test cases yet due to other personal priorities.

There is no rush. We all do open source coding on our spare time, so other life prioritize come first.

> Btw, based on the lang ref doc, I added defaults values whereever possible (as these entries might be helpful somewhere) when creating DataLayout struct, so the string I generate from that instance will have those extra spec entries which the layout string from the test data won't have, so direct string comparison will also not be possible, all we can do is create a datalayout instance from test data string, then generate a layout string, create another datalayout from this new string and use DeepEqual as @dannypsnl suggested. This is what I'll do, when I get time for it.

Sounds good!

> In the meantime, if you guys can add required enums in the other repo as @dannypsnl suggested, I'll update those here too.

Sure! We will see who gets to it first, @dannypsnl or me. I start with new University courses today, so will have less time on my hands for the next month or so.

Wish you both a good start of the early Autumn :)

Cheers,
Robin

svvivek commented:
Here is a sample test case

```go
func TestDataLayoutString(t *testing.T) {
	op_sys := "linux"
	arch := "x86-64"
	folderPath := "testdata/coreutils/test"
	os.Open("testdata")
	files, err := os.ReadDir(folderPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			t.Skip("testdata not found, clone it from https://github.com/llir/testdata")
		}
		t.Fatal(err.Error())
	}
	for _, f := range files {
		if !strings.HasSuffix(f.Name(), ".golden") {
			continue
		}
		filePath := folderPath + "/" + f.Name()
		t.Logf("=== [ %s ] ===", filePath)
		m, err := asm.ParseFile(filePath)
		if err != nil {
			t.Fatal(err.Error())
		}
		origDL, err := NewDataLayoutFromString(m.DataLayout, op_sys, arch)
		if err != nil {
			t.Error(err.Error())
		}
		newDL, err := NewDataLayoutFromString(origDL.LLString(), op_sys, arch)
		if err != nil {
			t.Error(err.Error())
		}
		if !reflect.DeepEqual(origDL, newDL) {
			t.Errorf("data layout mismatch, expected %q, got %q", origDL.LLString(), newDL.LLString())
		}
	}
}
```
The contents inside testdata/llvm/test/ folder is confusing me as there are lots of nested files and dirs there, so I used coreutils golden files. Let me know the appropriate files to test and the review on above code, I'll include this in a commit post that.

Note:- df.ll.golden ends up in the following panic while running the test cases
`panic: unable to parse integer "u0x1000"; strconv.ParseInt: parsing "u0x1000": invalid syntax`
I tried skipping that but multiple files return similar errors.

Btw, I also had to update go.mod to use llir/llvm v0.3.4 (currently it uses v0.3.3) for asm.ParseFile to parse correctly few cases updated to 12.0.1

mewmew commented:
> The contents inside testdata/llvm/test/ folder is confusing me as there are lots of nested files and dirs there, so I used coreutils golden files. Let me know the appropriate files to test and the review on above code, I'll include this in a commit post that.

Coreutil golden files are fine. They were produced on my laptop, so will probably only have x86-64, Linux data layout. The [testdata/llvm/test](https://github.com/llir/testdata/tree/master/llvm/test) is a copy of the official test cases of LLVM, that is every `*.ll` file of the official LLVM release. If you want, you can grep for `datalayout` in this directory to find example strings to use for test cases.

<details><summary>`datalayout` examples from `testdata/llvm/test/`</summary>

```llvm
target datalayout = "E-i64:64-n32:64-S128"
target datalayout = "E-m:e-i1:8:16-i8:8:16-i64:64-f128:64-a:8:16-n32:64"
target datalayout = "E-m:e-i1:8:16-i8:8:16-i64:64-f128:64-v128:64-a:8:16-n32:64"
target datalayout = "E-m:e-i64:64-i128:128-n32:64-S128"
target datalayout = "E-m:e-i64:64-n32:64"
target datalayout = "E-m:e-i64:64-n32:64"
target datalayout = "E-m:e-i64:64-n32:64-S128"
target datalayout = "E-m:e-i8:8:32-i16:16:32-i64:64-n32:64-S128"
target datalayout = "E-m:e-p:32:32-i128:64-n32"
target datalayout = "E-m:e-p:32:32-i64:32-f64:32:64-a:0:32-n32-S128"
target datalayout = "E-m:e-p:32:32-i64:64-a:0:32-n32-S64"
target datalayout = "E-m:e-p:32:32-i64:64-f128:64-n32-S64"
target datalayout = "E-m:e-p:32:32-i64:64-n32"
target datalayout = "E-m:e-p:32:32-i64:64-v128:64:128-n32-S64"
target datalayout = "E-m:e-p:32:32-i8:8:8-i16:16:16-i64:32:32-f64:32:32-v64:32:32-v128:32:32-a0:0:32-n32"
target datalayout = "E-m:m-i8:8:32-i16:16:32-i64:64-n32:64-S128"
target datalayout = "E-m:m-p:32:32-i8:8:32-i16:16:32-i64:64-n32-S64"
target datalayout = "E-m:o-p:32:32-f64:32:64-n32"
target datalayout = "E-p:32:32"
target datalayout = "E-p:32:32:32-S0-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f16:16:16-f32:32:32-f64:64:64-f128:64:128-v64:64:64-v128:128:128-a0:0:64-n32"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:32-i16:16:32-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-n32"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:32-i16:16:32-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-n32"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:32-i16:16:32-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-n32-S64"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f128:64:128"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f128:64:128-n32"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-f128:128:128"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v128:128:128-n32"
target datalayout = "E-p:32:32:32-i1:8:8-i8:8:8-i8:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f128:64:128-n32"
target datalayout = "E-p:64:64"
target datalayout = "E-p:64:64:64"
target datalayout = "E-p:64:64:64-S0-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f16:16:16-f32:32:32-f64:64:64-f128:128:128-v64:64:64-v128:128:128-a0:0:64-n32:64"
target datalayout = "E-p:64:64:64-a0:0:8-f32:32:32-f64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-v64:64:64-v128:128:128"
target datalayout = "E-p:64:64:64-a0:0:8-f32:32:32-f64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-v64:64:64-v128:128:128-n32:64"
target datalayout = "E-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f128:64:128"
target datalayout = "E-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "E-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n8:16:32:64"
target datalayout = "E-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-f128:128:128-v128:128:128-n32:64"
target datalayout = "E-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v128:128:128-n32:64"
target datalayout = "E-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-n32:64-S128"
target datalayout = "E-p:64:64:64-p1:16:16:16-a0:0:8-f32:32:32-f64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-v64:64:64-v128:128:128"
target datalayout = "E-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "E-p:64:64:64-p1:32:32:32-a0:0:8-f32:32:32-f64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-v64:64:64-v128:128:128"
target datalayout = "E-p:64:64:64-p1:64:64:64-p2:32:32:32-a0:0:8-f32:32:32-f64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-v64:64:64-v128:128:128"
target datalayout = "Fi24"
target datalayout = "G16777216"
target datalayout = "P1"
target datalayout = "P40"
target datalayout = "S24"
target datalayout = "e"
target datalayout = "e-P1-p:16:8-i8:8-i16:8-i32:8-i64:8-f32:8-f64:8-n8-a:8"
target datalayout = "e-P1-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128"
target datalayout = "e-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-p:32:32:32-v128:32:32"
target datalayout = "e-i32:32:64"
target datalayout = "e-i64:64-f80:128-n8:16:32:64"
target datalayout = "e-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-i64:64-f80:128-s:64-n8:16:32:64-S128"
target datalayout = "e-i64:64-i128:128-v16:16-v32:32-n16:32:64"
target datalayout = "e-i64:64-n32:64-S128"
target datalayout = "e-i64:64-v16:16-v32:32-n16:32:64"
target datalayout = "e-i64:64-v16:16-v32:32-n16:32:64-p:64:64:64-p1:32:32:32"
target datalayout = "e-m:e-i32:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-i32:64-f80:128-n8:16:32:64-S128-ni:1"
target datalayout = "e-m:e-i32:64-i128:128-n32:64-S128"
target datalayout = "e-m:e-i54:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:1"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:10:11:12"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:10:11:12:13"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:4"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:4:5"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-p100:128:64:64-p101:128:64:64"
target datalayout = "e-m:e-i64:64-i128:128-n32:64-S128"
target datalayout = "e-m:e-i64:64-i128:128-n32:64-S128-p:16:16-p4:32:16"
target datalayout = "e-m:e-i64:64-i128:128-n8:16:32:64-S128"
target datalayout = "e-m:e-i64:64-n32:64"
target datalayout = "e-m:e-i64:64-n32:64"
target datalayout = "e-m:e-i64:64-n32:64-v128:128:128"
target datalayout = "e-m:e-i64:64-n32:64-v256:256:256-v512:512:512"
target datalayout = "e-m:e-i64:64-n8:16:32:64"
target datalayout = "e-m:e-i64:64-p:64:64-n32:64-v256:256:256-v512:512:512"
target datalayout = "e-m:e-i64:64-p:64:64:64-n8:16:32:64-S128"
target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
target datalayout = "e-m:e-p200:32:32-p201:32:32-p202:64:64-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target datalayout = "e-m:e-p:16:16-i32:16-i64:16-f32:16-f64:16-a:8-n8:16-S16"
target datalayout = "e-m:e-p:16:16-i32:16:32-a:16-n8:16"
target datalayout = "e-m:e-p:16:16-p1:32:16-i32:16-i64:16-n8:16"
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64-ni:2"
target datalayout = "e-m:e-p:32:32-f64:32:64-f80:32-n8:16:32-S128"
target datalayout = "e-m:e-p:32:32-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32"
target datalayout = "e-m:e-p:32:32-i1:32-i64:64-a:0-n32"
target datalayout = "e-m:e-p:32:32-i1:32-i64:64-a:0-v32:32-n16:32"
target datalayout = "e-m:e-p:32:32-i1:8:32-i8:8:32-i16:16:32-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32"
target datalayout = "e-m:e-p:32:32-i1:8:32-i8:8:32-i16:16:32-i64:64-v128:64:128-a:0:32-n32-S64"
target datalayout = "e-m:e-p:32:32-i64:64-a:0-v32:32-n16:32"
target datalayout = "e-m:e-p:32:32-i64:64-f128:64-n32-S64"
target datalayout = "e-m:e-p:32:32-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-p:32:32-i64:64-n32-S128"
target datalayout = "e-m:e-p:32:32-i64:64-n32:64-S128"
target datalayout = "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
target datalayout = "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n8:16:32-S64"
target datalayout = "e-m:e-p:32:32-i64:64-v128:64:128-n32-S64"
target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-f64:32:64-f80:32-n8:16:32-S128"
target datalayout = "e-m:e-p:32:32:32-a:0-n16:32-i64:64:64-i32:32:32-i16:16:16-i1:8:8-f32:32:32-f64:64:64-v32:32:32-v64:64:64-v512:512:512-v1024:1024:1024-v2048:2048:2048"
target datalayout = "e-m:e-p:32:32:32-i64:64:64-i32:32:32-i16:16:16-i1:8:8-f64:64:64-f32:32:32-v64:64:64-v32:32:32-a:0-n16:32"
target datalayout = "e-m:e-p:64:64-i128:64-v128:64:128-a:0:64-n64-S64"
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target datalayout = "e-m:e-p:64:64-i64:64-n32:64-S128"
target datalayout = "e-m:e-p:64:64:64-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:e-p:64:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1"
target datalayout = "e-m:m-i8:8:32-i16:16:32-i64:64-n32:64-S128"
target datalayout = "e-m:m-p1:64:64:64-p:32:32:32-n8:16:32"
target datalayout = "e-m:m-p:40:64:64:32-i32:32-i16:16-i8:8-n32"
target datalayout = "e-m:o-i32:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128-ni:1"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128-p1:64:64"
target datalayout = "e-m:o-i64:64-f80:128-n8:8:32:64-S128"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:o-p:32:32-Fi8-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32"
target datalayout = "e-m:o-p:32:32-Fi8-i64:64-a:0:32-n32-S128"
target datalayout = "e-m:o-p:32:32-f64:32:64-f80:128-n8:16:32-S128"
target datalayout = "e-m:o-p:32:32-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32"
target datalayout = "e-m:o-p:32:32-i1:8:32-i8:8:32-i16:16:32-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32"
target datalayout = "e-m:o-p:32:32-i64:64-a:0:32-n32-S128"
target datalayout = "e-m:o-p:32:32-i64:64-i128:128-n32:64-S128"
target datalayout = "e-m:o-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
target datalayout = "e-m:o-p:40:64:64:32-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:o-p:64:64-f64:32:64-f80:128-n8:16:32-S128"
target datalayout = "e-m:w-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-m:w-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target datalayout = "e-m:w-p:32:32-i64:64-f80:32-n8:16:32-S32"
target datalayout = "e-m:w-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
target datalayout = "e-m:w-p:64:64-i32:32-i64:64-i128:128-n32:64-S128"
target datalayout = "e-m:x-p:32:32-i64:64-f80:32-n8:16:32-S32"
target datalayout = "e-m:x-p:32:32-i64:64-f80:32-n8:16:32-a:0:32-S32"
target datalayout = "e-m:x-p:32:32-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:32-n8:16:32-a:0:32-S32"
target datalayout = "e-n32"
target datalayout = "e-n32:64"
target datalayout = "e-p1:16:16-p2:32:32-p3:64:64"
target datalayout = "e-p3:32:32-p4:64:64-n32"
target datalayout = "e-p:128:128:128-p1:32:32:32-p2:8:8:8-p3:16:16:16-p4:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32"
target datalayout = "e-p:16:16"
target datalayout = "e-p:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:16:16"
target datalayout = "e-p:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:16:32"
target datalayout = "e-p:16:16:16-i8:8:8-i16:16:16-i32:16:16"
target datalayout = "e-p:16:16:16-i8:8:8-i16:16:16-i32:16:32-n8:16"
target datalayout = "e-p:16:16:16-i8:8:8-i16:16:16-i32:16:32-n8:16-a0:16:16"
target datalayout = "e-p:16:8:8-i8:8:8-i16:8:8-i32:8:8"
target datalayout = "e-p:16:8:8-i8:8:8-i16:8:8-i32:8:8-n8:16"
target datalayout = "e-p:16:8:8-i8:8:8-i8:8:8-i32:8:8"
target datalayout = "e-p:32:32"
target datalayout = "e-p:32:32-f64:32:64-i64:32:64-f80:32:32"
target datalayout = "e-p:32:32-i64:64-v16:16-v32:32-n16:32:64"
target datalayout = "e-p:32:32-n32"
target datalayout = "e-p:32:32-n32-S64"
target datalayout = "e-p:32:32-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
target datalayout = "e-p:32:32-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-A5"
target datalayout = "e-p:32:32-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-p24:64:64-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
target datalayout = "e-p:32:32-p1:64:64-p4:64:64"
target datalayout = "e-p:32:32-p4:64:64"
target datalayout = "e-p:32:32:32"
target datalayout = "e-p:32:32:32-S128-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f16:16:16-f32:32:32-f64:32:64-f128:128:128-v64:64:64-v128:128:128-a0:0:64-f80:32:32-n8:16:32"
target datalayout = "e-p:32:32:32-a0:0:32-n32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:32-f16:16:32-f32:32:32-f64:32:32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:32-f32:32:32-f64:32:32-v64:32:64-v128:32:128-a0:0:32-n32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:32-f32:32:32-f64:32:32-v64:64:64-v128:128:128-a0:0:32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:32-f32:32:32-f64:32:32-v64:64:64-v128:128:128-a0:0:32-n32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:32:64-v128:32:128-a0:0:32-n32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:32:64-v128:32:128-a0:0:32-n32-S32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:32:64-v128:32:128-v256:32:256-a0:0:32-n32-S32"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:64:128-a0:0:32-n32-S64"
target datalayout = "e-p:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:64:128-a0:0:32-n8:16:32-S64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:32-i16:16:32-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:32-i16:16:32-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-n32-S64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-f64:32:64-v64:64:64-v128:128:128"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32-f32:32:32-f64:32:32-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32-f32:32:32-f64:32:32-v64:64:64-v128:128:128-a0:0:64-n32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:128:128-v64:64:64-v128:128:128-a0:0:64-f80:128:128-n8:16:32-S128"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-f80:32:32-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:32:64-v128:32:128-a0:0:32-n32-S32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:32:64-v128:32:128-a0:0:64-n32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128-f128:128:128-n8:16:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128-n32:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128-n8:16:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128-n8:16:32-S128"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128:n8:16:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:32:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:32:32-n8:16:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:32:32-n8:16:32-S128"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-n8:16:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-s0:0:64-f80:32:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-f80:128:128-v64:64:64-v128:128:128-a0:0:64-f80:32:32-n8:16:32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-f80:128:128-v64:64:64-v128:128:128-a0:0:64-f80:32:32-n8:16:32-S32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-n32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-f80:32:32-n8:16:32-S32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n32-S64"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:64:128-a0:0:64-n32"
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:64:128-a0:0:64-n32-S64"
target datalayout = "e-p:32:32:32-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v24:32:32-v32:32:32-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:32:32:32-i64:64:64-i32:32:32-i16:16:16-i1:32:32-f64:64:64-f32:32:32-a0:0-n32"
target datalayout = "e-p:32:32:32-i64:64:64-i32:32:32-i16:16:16-i1:32:32-f64:64:64-f32:32:32-v64:64:64-v32:32:32-a0:0-n16:32"
target datalayout = "e-p:32:32:32-i64:64:64-i32:32:32-i16:16:16-i1:32:32-f64:64:64-f32:32:32-v64:64:64-v32:32:32-a:0-n16:32"
target datalayout = "e-p:32:32:32-n32"
target datalayout = "e-p:32:32:32-n8:16:32"
target datalayout = "e-p:32:32:32-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-n8:16:32-S128"
target datalayout = "e-p:32:32:32-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128"
target datalayout = "e-p:32:32:32-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-v16:16:16-v24:32:32-v32:32:32-v48:64:64-v64:64:64-v96:128:128-v128:128:128-v192:256:256-v256:256:256-v512:512:512-v1024:1024:1024-v2048:2048:2048-n32:64"
target datalayout = "e-p:32:32:32-p1:16:16:16-p2:32:32:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:32:64-v128:32:128-a0:0:32-n8:16:32"
target datalayout = "e-p:32:32:32-p1:16:16:16-p2:8:8:8-p4:64:64:64-n16:32:64"
target datalayout = "e-p:32:32:32-p1:64:64:64-p2:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32-n8:16:32:64"
target datalayout = "e-p:32:32:32-p1:64:64:64-p2:8:8:8-p3:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32"
target datalayout = "e-p:32:32:32-p1:64:64:64-p2:8:8:8-p3:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32-n8:16:32:64"
target datalayout = "e-p:32:32:32-p1:64:64:64-p2:8:8:8-p3:16:16:16-p4:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:32"
target datalayout = "e-p:40:64:64:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:40:64:64:32-p1:16:16:16-p2:32:32:32-p3:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:64:32-i64:32-v32:32-n32-S64"
target datalayout = "e-p:64:64"
target datalayout = "e-p:64:64-i32:32:32-i64:64:64-f32:32:32-f64:64:64"
target datalayout = "e-p:64:64-i64:64-f80:128-n8:16:32:64-S128"
target datalayout = "e-p:64:64-i64:64-n32:64"
target datalayout = "e-p:64:64-p1:16:16-p2:32:32:32-p3:64:64:64"
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-A5"
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5"
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-ni:7"
target datalayout = "e-p:64:64-p1:64:64-p2:64:64-p3:32:32-p4:32:32-p5:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-A5"
target datalayout = "e-p:64:64-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-A5"
target datalayout = "e-p:64:64-p1:64:64-p5:32:32"
target datalayout = "e-p:64:64-p5:32:32-A5"
target datalayout = "e-p:64:64:64"
target datalayout = "e-p:64:64:64-S128-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f16:16:16-f32:32:32-f64:64:64-f128:128:128-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:64:64:64-S128-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-f128:128:128-n8:16:32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n8:16:32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:32:32-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v128:128:128-n32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n32:64-S128"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-f128:128:128-n8:16:32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128:n8:16:32:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-n8:16:32:64-S128"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-i128:128:128-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128-p100:128:64:64-p101:128:64:64"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i8:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128"
target datalayout = "e-p:64:64:64-n32:64"
target datalayout = "e-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-f80:128-v64:64:64-v128:128:128-a0:0:64-n8:16:32:64"
target datalayout = "e-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n8:16:32:64"
target datalayout = "e-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128"
target datalayout = "e-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128"
target datalayout = "e-p:64:64:64-p1:16:16:16-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-n8:16:32:64-S128"
target datalayout = "e-p:64:64:64-p1:16:16:16-n16:32:64"
target datalayout = "e-p:64:64:64-p1:16:16:16-p2:32:32-p3:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n8:16:32:64-A2"
target datalayout = "e-p:64:64:64-p1:16:16:16-p2:32:32:32-p3:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "e-p:64:64:64-p1:16:16:16-p3:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-n8:16:32:64"
target datalayout = "e-p:64:64:64-p1:32:32:32-p2:16:16:16-n8:16:32:64"
target datalayout = "e-p:64:64:64:32"
target datalayout = "e-p:64:64:64:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
target datalayout = "f64:64:64-v64:64:64"
target datalayout = "i32:8:32"
target datalayout = "i32:8:8"
target datalayout = "m:o"
target datalayout = "m:w"
target datalayout = "n16"
target datalayout = "n2:8:16:32:64"
target datalayout = "n32"
target datalayout = "n32:64"
target datalayout = "n64"
target datalayout = "n8:16:32:64"
target datalayout = "n8:32"
target datalayout = "p32:32:32-p1:16:16:16-n16:32"
target datalayout = "p:16:16"
target datalayout = "p:32:32"
target datalayout = "p:32:32-p1:32:32-p2:16:16"
target datalayout = "p:32:32-p1:64:64"
target datalayout = "p:32:32-p1:64:64-p2:16:16-n16:32:64"
target datalayout = "p:32:32-p3:32:32-p4:64:64"
target datalayout = "p:32:32:32"
target datalayout = "p:32:32:32-p1:16:16:16"
target datalayout = "p:32:32:32-p1:32:32:32-p2:16:16:16"
target datalayout = "p:40:64:64:32"
target datalayout = "p:64:64:64"
target datalayout = "p:64:64:64-i64:32:32"
target datalayout = "p:64:64:64-n32:64"
target datalayout = "p:8:8-n8"
```

</details>

I'm not sure if all of the examples above are valid strings, perhaps some are used in test cases of LLVM for invalid strings. Just as a notice.

mewmew commented:
> Btw, I also had to update go.mod to use llir/llvm v0.3.4 (currently it uses v0.3.3) for asm.ParseFile to parse correctly few cases updated to 12.0.1

Thanks for reporting. Updates to the latest version in 6e37aa734e1b0cca82ebce2ceeecffec14de6e9e.

mewmew commented:
> Note:- df.ll.golden ends up in the following panic while running the test cases
panic: unable to parse integer "u0x1000"; strconv.ParseInt: parsing "u0x1000": invalid syntax
I tried skipping that but multiple files return similar errors.

Hmm, will have to investigate. @dannypsnl do you know what causes this?

mewmew commented:
> Here is a sample test case

@svvivek That's perfect! Really great test case with round trip decoding and encoding.

svvivek commented:
> > Here is a sample test case
> 
> @svvivek That's perfect! Really great test case with round trip decoding and encoding.

Shall I submit the testcase for golden utils alone now and add the llvm/test related later when I get time ? If yes, we can close this PR and add the extra test cases later.

svvivek commented:
> > > Here is a sample test case
> > 
> > 
> > @svvivek That's perfect! Really great test case with round trip decoding and encoding.
> 
> Shall I submit the testcase for golden utils alone now and add the llvm/test related later when I get time ? If yes, we can close this PR and add the extra test cases later.

Reason being, some default values I added are only for x86-64 and linux as you would've seen in the PR, so testing other os/arch combinations would mean I spend sometime updating the default values which are not actually there in lang ref doc and I have to search online.

svvivek committed:
Merge branch 'llir:master' into master

dannypsnl commented:
> > Note:- df.ll.golden ends up in the following panic while running the test cases
> > panic: unable to parse integer "u0x1000"; strconv.ParseInt: parsing "u0x1000": invalid syntax
> > I tried skipping that but multiple files return similar errors.
> 
> Hmm, will have to investigate. @dannypsnl do you know what causes this?

Read https://pkg.go.dev/strconv#ParseInt, can see `0x1000` is valid when `base = 0`. But `u0x1000` is invalid in any condition, we have to manually handle LLVM number string like we do in `llir/llvm` repo.

mewmew commented:
> >Shall I submit the testcase for golden utils alone now and add the llvm/test related later when I get time ? If yes, we can close this PR and add the extra test cases later.
>
>Reason being, some default values I added are only for x86-64 and linux as you would've seen in the PR, so testing other os/arch combinations would mean I spend sometime updating the default values which are not actually there in lang ref doc and I have to search online.

Sounds good to me. Lets add the golden test cases first using `testdata/coreutils`, then we can use `testdata/llvm/test` later when those default values are implemented.

dannypsnl commented:
@mewmew we can reuse [this](https://github.com/llir/llvm/blob/master/ir/constant/const_int.go#L47-L102) here.

mewmew commented:
> Note:- df.ll.golden ends up in the following panic while running the test cases
panic: unable to parse integer "u0x1000"; strconv.ParseInt: parsing "u0x1000": invalid syntax
I tried skipping that but multiple files return similar errors.

@svvivek could you clarify where you get this panic? Running `go test ./...` on the latest version of `llir/llvm` gives a PASS and it parses `testdata/coreutils/test/df.ll` among others.

svvivek commented:
> > Note:- df.ll.golden ends up in the following panic while running the test cases
> > panic: unable to parse integer "u0x1000"; strconv.ParseInt: parsing "u0x1000": invalid syntax
> > I tried skipping that but multiple files return similar errors.
> 
> @svvivek could you clarify where you get this panic? Running `go test ./...` on the latest version of `llir/llvm` gives a PASS and it parses `testdata/coreutils/test/df.ll` among others.

I have committed the testcase. Just do **go test -v .** in this repo and you will see the panic.

svvivek committed:
- Testcase added for DataLayout. Ref - PR #4

mewmew commented:
> I have committed the testcase. Just do go test -v . in this repo and you will see the panic.

Perfect! Thanks for the reproducible test case. The issue has now been fixed (in rev llir/llvm@252431239101c672e2021de2d53f92a1443c524d).

Cheers,
Robin

svvivek commented:
I hope everything is resolved, can we merge this PR or is there anything else to do @mewmew ?

mewmew commented:
> I hope everything is resolved, can we merge this PR or is there anything else to do @mewmew ?

Looks great! Thanks for working on this @svvivek!

Cheers,
Robin
